From 0a7ff2ca42cfe2e939248b0dc1fdfc756c0b4133 Mon Sep 17 00:00:00 2001
From: Nikolaus Rath <Nikolaus@rath.org>
Date: Thu, 8 Oct 2015 12:24:37 -0700
Subject: Work around bug in mips+mipsel libc

Forwarded: no
Last-Update: 2015-09-16

On mips and mipsel, the st_dev and st_rdev members of struct stat do not
have type dev_t. This breaks POSIX compatibility, but is difficult to fix
(cf. https://sourceware.org/bugzilla/show_bug.cgi?id=17786).

To work around the issue, we change the definition of struct stat that
is used by Cython when we are compiling under mips. Note that this
requires the Cython compilation to run under mips, and that the
resulting C file will be mips specific (without the patch, the
generated C file is suitable for any architecture).

Upstream is not interested in this change for obvious reasons.

Patch-Name: mips_dev_t.diff
---
 Include/libc/sys/stat.pxd | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/Include/libc/sys/stat.pxd b/Include/libc/sys/stat.pxd
index 2df8173..7506d78 100644
--- a/Include/libc/sys/stat.pxd
+++ b/Include/libc/sys/stat.pxd
@@ -16,7 +16,25 @@ cdef extern from "sys/stat.h" nogil:
         S_IFMT
         S_IFDIR
 
-    struct stat:
+IF UNAME_MACHINE.startswith('mips64'):
+    cdef extern from "sys/stat.h" nogil:
+      struct stat:
+        int       st_dev
+        ino_t     st_ino
+        mode_t    st_mode
+        nlink_t   st_nlink
+        uid_t     st_uid
+        gid_t     st_gid
+        int       st_rdev
+        off_t     st_size
+        blksize_t st_blksize
+        blkcnt_t  st_blocks
+        time_t    st_atime
+        time_t    st_mtime
+        time_t    st_ctime
+ELSE:
+    cdef extern from "sys/stat.h" nogil:
+      struct stat:
         dev_t     st_dev
         ino_t     st_ino
         mode_t    st_mode
